File: vim/functions/plug_get_latest_commits.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

def execute_in_vim_dir(&block)
  vim_plugs = `ls ~/.vim/plugged`.split("\n")
  vim_plugs.each do |dir|
    Dir.chdir("#{Dir.home}/.vim/plugged/#{dir}", &block)
  end
end

def repo_name
  remote = `git remote get-url origin`.strip
  remote_url = remote.sub(/\.git\z/, '')
  remote_url.gsub('https://github.com/', '')
end

def commit
  `git rev-parse HEAD`.strip
end

execute_in_vim_dir do
  case repo_name
  when 'dracula/vim'
    puts "Plug '#{repo_name}', { 'commit': '#{commit}', 'as': 'dracula' }"
  when 'masakiq/vim-ruby-fold'
    puts "Plug '#{repo_name}', { 'commit': '#{commit}', 'for': 'ruby' }"
  when 'dart-lang/dart-vim-plugin'
    puts "Plug '#{repo_name}', { 'commit': '#{commit}', 'for': 'dart' }"
  when 'maxmellon/vim-jsx-pretty'
    puts "Plug '#{repo_name}', { 'commit': '#{commit}', 'for': ['javascript', 'typescript'] }"
  when 'rcmdnk/vim-markdown'
    puts "Plug '#{repo_name}', { 'commit': '#{commit}', 'for': 'markdown' }"
  when 'iamcco/markdown-preview.nvim', 'masakiq/vim-markdown-composer'
    # ignore
  else
    puts "Plug '#{repo_name}', { 'commit': '#{commit}' }"
  end
end

File: vim/lua_scripts/plug_get_latest_commits.lua
#!/usr/bin/env lua

local function execute_shell_command(cmd)
  local handle = io.popen(cmd)
  if not handle then
    return nil
  end
  local result = handle:read("*a")
  handle:close()
  return result
end

local function run_in_vim_plugin_dirs(callback)
  local vim_plugin_dirs = execute_shell_command('ls ~/.vim/plugged')
  if not vim_plugin_dirs then
    return
  end
  for dir in vim_plugin_dirs:gmatch("[^\r\n]+") do -- split by newline
    local path = "~/.vim/plugged/" .. dir
    local cmd = string.format('cd %s && lua -e "%s"', path, callback)
    os.execute(cmd)
  end
end

local callback_code = [[
  function repo_name()
    local handle = io.popen('git remote get-url origin')
    local remote = handle:read('*a'):gsub('%s+', '')
    handle:close()

    local remote_url = remote:gsub('%.git$', '')
    return remote_url:gsub('https://github.com/', '')
  end

  function commit()
    local handle = io.popen('git rev-parse HEAD')
    local commit_id = handle:read('*a'):gsub('%s+', '')
    handle:close()

    return commit_id
  end

  local repo = repo_name()
  local commit_id = commit()
  local plugin_declaration = string.format('Plug \'%s\', { \'commit\': \'%s\'', repo, commit_id)

  if repo == 'dracula/vim' then
    print(plugin_declaration .. ', \'as\': \'dracula\' }')
  elseif repo == 'masakiq/vim-ruby-fold' then
    print(plugin_declaration .. ', \'for\': \'ruby\' }')
  elseif repo == 'dart-lang/dart-vim-plugin' then
    print(plugin_declaration .. ', \'for\': \'dart\' }')
  elseif repo == 'maxmellon/vim-jsx-pretty' then
    print(plugin_declaration .. ', \'for\': [\'javascript\', \'typescript\'] }')
  elseif repo == 'rcmdnk/vim-markdown' then
    print(plugin_declaration .. ', \'for\': \'markdown\' }')
  elseif repo == 'neoclide/coc.nvim' then
    print(plugin_declaration .. ', \'do\': \'npm ci\' }')
  elseif repo == 'iamcco/markdown-preview.nvim' then
    -- ignore
  else
    print(plugin_declaration .. ' }')
  end
]]

run_in_vim_plugin_dirs(callback_code)
